lib/vmpooler/pool_manager.rb

```diff
@@ -319,10 +319,62 @@ def _destroy_vm(vm, pool, provider)
         finish = format('%.2f', Time.now - start)
         $logger.log('s', "[-] [#{pool}] '#{vm}' destroyed in #{finish} seconds")
         $metrics.timing("destroy.#{pool}", finish)
+        get_vm_usage_labels(vm)
       end
       dereference_mutex(vm)
     end
 
+    def get_vm_usage_labels(vm)
+      return unless $config[:config]['usage_stats']
+      checkout = $redis.hget("vmpooler__vm__#{vm}", 'checkout')
+      return if checkout.nil?
+      jenkins_build_url = $redis.hget("vmpooler__vm__#{vm}", 'tag:jenkins_build_url')
+      user = $redis.hget("vmpooler__vm__#{vm}", 'token:user') || 'unauthenticated'
+      poolname = $redis.hget("vmpooler__vm__#{vm}", "template")
+
+      if jenkins_build_url
+        url_parts = jenkins_build_url.split('/')[2..-1]
+        instance = url_parts[0].gsub('.', '_')
+        value_stream_parts = url_parts[2].split('_')
+        value_stream = value_stream_parts.shift
+        branch = value_stream_parts.pop
+        project = value_stream_parts.shift
+        job_name = value_stream_parts.join('_')
+        build_metadata_parts = url_parts[3]
```

Comment:
I think this could be `nil` if the job is not a matrix. Then we will hit a NilClass error when trying to split it on line 368
```
 labels_string_parts = labels_string.split(',')
```

The tests are great, please add one for a job that doesnt have that last part eg `https://host/job/jobname/` or does ABS alway also include the build number like `https://host/job/jobname/220` in which case I guess we wouldn't get the `nil` error but rather parse `220` in `component_to_test()`

That being said we should also think about error handling. Probably errors thrown by this whole method could be handled instead of throwing a stacktrace or an error in the log if this is wrapped. I would rather have us ignore the error and keep whatever processing the destroy should be doing (is there much left in that logic path?)

Reply:
This is the second to last step in the destroy. The only other thing that happens is the mutex is dereferenced so it can be removed during garbage collection. It probably is worth capturing any failures and preventing them from bubbling up, and perhaps adding a log message for when parsing the URL fails.

Reply:
I added a test to cover a jenkins job URL with no matrix information. I think you're correct that the build number ends up getting passed instead of nil. In any case, I added test coverage for both examples so we should be able to handle these cases without an error.